### ISSUE TITLE
many tests fail on Linux: "Too many open files"

### ISSUE THREAD

**Issue by strager** [facebook/sapling#464]
Hundreds of tests are failing on my Linux machine with errors like:
```
  abort: Too many open files: $TESTTMP/repo/.sl/merge/state                    

  abort: internal storage is corrupted                                         
    "$TESTTMP/repo/.sl/blackbox/v1": cannot read 'latest'                      
    in rotate::OpenOptions::open("$TESTTMP/repo/.sl/blackbox/v1")              
    Caused by 1 errors:                                                        
    - Too many open files (os error 24)                                        

  #   File "/usr/lib/python3.8/shutil.py", line 627, in _rmtree_safe_fd
  #     with os.scandir(topfd) as scandir_it:
  # OSError: [Errno 24] Too many open files: '$TESTTMP/t/.sl/blackbox/v1/0'
```

**Comment by strager** [comment]
I debugged it a bit. At the very least, some `IndexedLogHgIdDataStore` objects are leaked. These are `Arc`d, so I have a few theories:

* There's a reference cycle (probably in some higher-level components like `FileStore`).
* Python or Rust code is storing a reference to the `WorkingCopy` object in a global variable.

**Comment by strager** [comment]
On my machine, `ulimit -n` reports `1024`. I didn't test a higher limit.

**Comment by strager** [comment]
In `eden/scm/edenscm/testing/ext/hg.py`, if I set `inprocesshg = False`, the errors no longer happen (at least in the one test I tried).

**Comment by akushner** [comment]
Hey Matt! Thanks for all the issues reported and PRs.

While there might be a leak, it's worth noting that all of our internal laptops are configured by IT to have 1048576 open files.  Probably doesn't need to be that many, but 1024 is pretty minimalistic and  a number of internal builds needed more than 1024.

**Comment by metalim** [comment]
> While there might be a leak, it's worth noting that all of our internal laptops are configured by IT to have 1048576 open files. Probably doesn't need to be that many, but 1024 is pretty minimalistic and a number of internal builds needed more than 1024.

macOS has 256 max open files by default, all other apps are happy with the limit. Have you tried - I don't know - closing files?
It'll be a bummer if SCM is abandoned by community just because it wastes resources

**Comment by akushner** [comment]
@metalim, @strager received the EMFILE (errno 24) when running the highly parallelized test harness, not while running sl as a user.  The test harness defaults to running multiprocessing.cpu_count() (e.g 10 on my laptop) test jobs at a time which is likely why he saw this issue. In addition, the test runner defaults to using an in-process `sl`  service so all the open files of the tests were accumulated in one process.

If you hit this issue while running a command as a user, please let us know!

**Comment by vegerot** [comment]
> It'll be a bummer if SCM is abandoned by community just because it wastes resources

If there is a leak it's just during tests.  The way that the sapling tests work is by using one sl process and feeding many commands into it.  This helps improve test times.  All the FDs are closed once the sl process finishes, and so this retention is unlikely to be a problem when using sl normally.

I do agree that asking OSS users to increase their ulimit to run Sapling's test suite is inconvenient and should probably be fixed for the OSS build

**Comment by metalim** [comment]
> If there is a leak it's just during tests.

I've hit the MacOS limit consistently by just cloning the repo with 4 submodules. Didn't run the tests

**Comment by vegerot** [comment]
> > If there is a leak it's just during tests.
> 
> I've hit the limit on MacOS consistently by just cloning the repo with 4 submodules. Didn't run the tests

Now _that's_ a problem.  Would you please provide a bit more information?  What command did you run?  Just 

```sh
sl clone git@github.com:facebook/sapling.git
```
?
Would you please provide more information about your macOS version and Sapling version?  What is your ulimit set to?

**Comment by metalim** [comment]
> What command did you run?

`sl clone git@github.com:metalim/sites
` -- that's small private repo with 4 submodules, also small


> macOS version and Sapling version?

Ventura 13.2
Sapling 0.2.20230124-180750  (from Homebrew)


> What is your ulimit set to?

```
~ using ☁️ saml
➜ ulimit -n
256

~ using ☁️ saml
➜ sudo launchctl limit maxfiles
	maxfiles    256            unlimited

```

Crashlog:

```
➜ sl clone git@github.com:metalim/sites
remote: Enumerating objects: 58, done.
remote: Counting objects: 100% (58/58), done.
remote: Compressing objects: 100% (43/43), done.
remote: Total 58 (delta 12), reused 52 (delta 6), pack-reused 0
Unpacking objects: 100% (58/58), 942.99 KiB | 608.00 KiB/s, done.
From ssh://github.com/metalim/sites
 * [new ref]         d87f6183afc74086704791c794023896410aa3bf -> remote/main
pulling submodule mem.ee
** Sapling SCM (version 0.2.20230124-180750) has crashed:
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/__init__.py", line 106, in run
    dispatch.run(args, fin, fout, ferr, config)
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/dispatch.py", line 161, in run
    status = (dispatch(req) or 0) & 255
              ^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/dispatch.py", line 503, in dispatch
    ret = _runcatch(req)
          ^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/dispatch.py", line 712, in _runcatch
    return _callcatch(ui, req, _runcatchfunc)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/dispatch.py", line 721, in _callcatch
    return scmutil.callcatch(ui, req, func)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/scmutil.py", line 147, in callcatch
    return func()
           ^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/dispatch.py", line 683, in _runcatchfunc
    return _dispatch(req)
           ^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/dispatch.py", line 1218, in _dispatch
    ret = runcommand(
          ^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/ext/undo.py", line 104, in _runcommandwrapper
    return orig(lui, repo, cmd, fullargs, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/ext/journal.py", line 80, in runcommand
    return orig(lui, repo, cmd, fullargs, *args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/ext/sparse.py", line 504, in _tracktelemetry
    res = runcommand(lui, repo, *args)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/ext/copytrace.py", line 164, in _runcommand
    return orig(lui, repo, cmd, fullargs, ui, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/dispatch.py", line 913, in runcommand
    ret = _runcommand(ui, options, cmd, d)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/dispatch.py", line 1259, in _runcommand
    return cmdfunc()
           ^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/dispatch.py", line 1217, in <lambda>
    d = lambda: util.checksignature(func)(ui, *args, **strcmdopt)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/util.py", line 1295, in check
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/util.py", line 1295, in check
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/ext/sparse.py", line 595, in _clonesparsecmd
    return orig(ui, repo, *args, **opts)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/util.py", line 1295, in check
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/util.py", line 1295, in check
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/ext/remotefilelog/__init__.py", line 325, in cloneshallow
    return orig(ui, source, *args, **opts)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/util.py", line 1295, in check
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/commands/__init__.py", line 1608, in clone
    r = git.clone(ui, giturl, dest, update)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/git.py", line 180, in clone
    hg.updatetotally(repo.ui, repo, node, None)
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/hg.py", line 972, in updatetotally
    hasunresolved = _update(repo, checkout, updatecheck=updatecheck)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/hg.py", line 900, in update
    stats = updaterepo(repo, node, False, updatecheck=updatecheck)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/ext/remotenames.py", line 328, in exupdate
    res = orig(repo, *args, **opts)
          ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/hg.py", line 885, in updaterepo
    return mergemod.update(
           ^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/ext/sparse.py", line 419, in _update
    results = orig(repo, node, branchmerge, *args, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/util.py", line 4166, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/merge.py", line 2456, in update
    git.submodulecheckout(p2, matcher, force=force)
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/git.py", line 725, in submodulecheckout
    submod.checkout(node, force=force)
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/git.py", line 856, in checkout
    self.pullnode(repo, node)
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/git.py", line 845, in pullnode
    pullrefspecs(repo, self.url, [refspec])
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/git.py", line 571, in pullrefspecs
    _syncfromgit(repo)
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/git.py", line 529, in _syncfromgit
    repo.invalidate(clearfilecache=True)
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/localrepo.py", line 2160, in invalidate
    self.invalidatechangelog()
  File "/opt/homebrew/Cellar/python@3.11/3.11.1/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/edenscm/localrepo.py", line 2192, in invalidatechangelog
    self._rsrepo.invalidatechangelog()
error.RustError: "/Users/metalim/Sites/metalim/test/sites/mem.ee/.sl/../../.sl/store/gitmodules/0db5511acec299772bced4fee9ac13711eaac887/.sl/store/segments/v1/idmap2/rlock": cannot open for locking
in Log:try_clone
  Log.dir = SharedMeta { path: Filesystem("/Users/metalim/Sites/metalim/test/sites/mem.ee/.sl/../../.sl/store/gitmodules/0db5511acec299772bced4fee9ac13711eaac887/.sl/store/segments/v1/idmap2"), meta: Mutex { data: LogMetadata { primary_len: 12, indexes: {}, epoch: 2245251737117852400 }, poisoned: false, .. } }
Caused by 1 errors:
- Too many open files (os error 24)
```

**Comment by metalim** [comment]
As it is not related to tests, I've opened a new issue: https://github.com/facebook/sapling/issues/527

**Comment by zzl0** [comment]
Thanks @strager for reporting and debugging the issue. We recommend changing the open files limit [1] as a short-term solution (as mentioned by @akushner, we are setting open files limit to 1048576 internally).

```
echo "ulimit -n 1048576" >> ~/.bash_profile
echo "ulimit -n 1048576" >> ~/.zshrc
```

@metalim yes, this can happen when cloning a larger repositories. Thanks for using Sapling, and I apologize for this issue. Please use the settings mentioned above to unblock yourself. We will debug and fix the root cause of the problem. Unfortunately, I cannot provide an ETA at this time.

@vegerot thanks for helping address the issue.

[1] https://sapling-scm.com/docs/introduction/installation

**Comment by metalim** [comment]
> yes, this can happen when cloning a larger repositories

The thing is: my repo is really small

```
➜ find .|wc -l
     186

➜ du -hsx
 37M	.
```
